### PR TITLE
flat

### LINKED ISSUE #16534: flat

### PR DESCRIPTION
**Pull request by Faiq2001**
Close #16534

### CODE REVIEW

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py
Comment (xoiga123) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:127]: A decorator should be right above a function definition, please move it to the line before `def test_numpy_ndarray_property_flat`

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py
Comment (xoiga123) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: Please use helpers to generate the argument `item` in the decorator `@given` as well, since you are also using `item` to call the property `.flat`. Thank you 😄

**Review by xoiga123 — CHANGES_REQUESTED** [review]
Well done! The implementation and the test call is correct, now you just have to polish it a little bit to make it run and it'll be good to merge!

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py
Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: @given(
    dtype_x=helpers.dtype_and_values(
        available_dtypes=helpers.get_dtypes("valid", prune_function=False),
        ret_shape=True,
    ),
    item=helpers.item(
        available_items = helpers.get_items("valid", prune_function=False), 
        ret_shape=True,
    ),
)    
Can you see if this code is correct?

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: @given(
    dtype_x=helpers.dtype_and_values(
        available_dtypes=helpers.get_dtypes("valid", prune_function=False),
        ret_shape=True,
    ),
    item=helpers.item(
        available_items = helpers.ints(1,dtype_x[3]),
        ret_shape=True,
    ),
)    
What about it now?

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: image

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: This is the function that i used....it looks fine to me but wanted a confirmation from you...sorry to disturb you again and again

Comment (xoiga123) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: You can check this section on [Setting up Pytest](https://unify.ai/docs/ivy/overview/contributing/setting_up.html#setting-up-testing-in-pycharm)

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: image

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: What about this?

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:131]: > You can check this section on [Setting up Pytest](https://unify.ai/docs/ivy/overview/contributing/setting_up.html#setting-up-testing-in-pycharm)

I am not able to implement this. I am using VS Code and after setting up the environment I am confused how should i run local tests

Comment (xoiga123) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py]: @Faiq2001 to run frontend tests, try these command
```bash
# run all the tests in a file
pytest ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py
# run a specific test, which is the function you're implementing
pytest ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py::test_numpy_ndarray_property_flat
```
The general format is
```bash
pytest test_file_path::test_fn_name
```

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py]: @xoiga123 why am i getting errors while running these codes.
image

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py]: I resolved some of the errors which were coming when importing 'config' from setuptools in __init__.py. But the errors just keep on increasing

Comment (xoiga123) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py]: @Faiq2001 Sorry I can't merge this unless it's all good to go :mask:. It's fine if you have implemented the test successfully without setting up locally and verifying it, but currently the tests are not runnable on my end. It takes a little time to debug around and set things up but they will hugely save you time in the long run. You can check the [vscode setup channel](https://discord.com/channels/799879767196958751/1029037632245739610) if you need help :+1:, sorry that I won't be able to help much with vscode because I use pycharm :sob: But I think the docs are great for this.

Comment (xoiga123) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py]: > Also dont you think the following helper function is correct? Item takes integers randomly between 1 to size(which is the total number of elements).
> 
> > image

Please try to generate arguments in the `@given` decorator only. You can watch this fantastic [Ivy Tests youtube video](https://www.youtube.com/watch?v=2AwWuHIe2h8&list=PLwNuX3xB_tv9NPbAVHs0dLe3GQkR51vG9&index=16) to get a hang of things and how to write a test properly :muscle:

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py]: image
Code
image
Tester
Changed the tester function and the code function as well, but still not able to run the tests locally. Can you guide me as to where i am making a mistake

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py]: @xoiga123 I think i somehow unknowingly requested someone. It doesnt matter noh?

**Review by xoiga123 — CHANGES_REQUESTED** [review]
Sorry, not sure how this wound up in the Pending status :sweat:

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py
Comment (xoiga123) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:133]: The way you're testing this is wrong, as `check_equal` are only used for non tensor inputs, i.e. int, float, list, tuple...

You should be doing this:
```python
flat_ivy = x.flat
flat_generated = data[0].flatten()
flat_ivy = helpers.flatten_and_to_np(ret=flat_ivy)
flat_generated = helpers.flatten_and_to_np(ret=flat_generated)

assert np.allclose(flat_ivy, flat_generated)
```

**Review by xoiga123 — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py
Comment (xoiga123) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:133]: Now that I coded this out, I realized this is wrong and defeats the entire purpose of testing for `flatten` :sweat_smile: Please find another helper/way to convert them into numpy and then use `np.allclose` without using `flatten_and_to_np`

Comment (Faiq2001) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py]: > 
@xoiga123 since i have converted the arrays into a flat array so we can check it using check_equal noh?(they arent tensor any more)...moreover, we need to return a 1d iterator over the array elements noh?

File: .idea/ivy.iml
Comment (xoiga123) [inline .idea/ivy.iml:5]: Please revert this

**Pull request merged**
